exonum-java-binding/core/rust/integration_tests/tests/runtime_error_handling.rs

```diff
@@ -0,0 +1,194 @@
+        env.throw_new(ILLEGAL_ARGUMENT_EXCEPTION_CLASS, EXPECTED_DESCRIPTION)?;
+        wait_for_thrown_exception(env)?;
+        Ok(())
+    });
+}
+
+#[test]
+fn jni_call_transaction_transaction_execution_exception() {
+    let expected = ErrorMatch::from_fail(&FakeServiceError::Error)
+        .with_description_containing(EXPECTED_DESCRIPTION);
+    let err = jni_call_transaction(&EXECUTOR, |env| {
+        let exception = create_transaction_execution_exception(env)?;
+        env.throw(JThrowable::from(exception))?;
+        wait_for_thrown_exception(env)?;
+        Ok(())
+    })
+    .unwrap_err();
+    assert_eq!(err, expected);
+}
+
+#[test]
+fn jni_call_transaction_unexpected_transaction_execution_exception() {
+    let err = jni_call_transaction(&EXECUTOR, |env| {
+        let cause = create_any_java_exception(env)?;
+        let exception = env.new_object(
+            UNEXPECTED_TRANSACTION_EXECUTION_EXCEPTION_CLASS,
+            "(Ljava/lang/Throwable;)V",
+            &[cause.into()],
+        )?;
+        env.throw(JThrowable::from(exception))?;
+        wait_for_thrown_exception(env)?;
+        Ok(())
+    })
+    .unwrap_err();
+    assert_eq!(err.kind(), ErrorKind::Unexpected);
+    assert_eq!(err.description(), EXPECTED_DESCRIPTION);
+}
+
+fn create_transaction_execution_exception<'a>(env: &'a JNIEnv) -> JniResult<JObject<'a>> {
+    let code = EXPECTED_CODE;
+    let description: JObject = env.new_string(EXPECTED_DESCRIPTION)?.into();
+    env.new_object(
+        TRANSACTION_EXECUTION_EXCEPTION_CLASS,
+        "(BLjava/lang/String;)V",
+        &[code.into(), description.into()],
+    )
+}
+
+fn create_any_java_exception<'a>(env: &'a JNIEnv) -> JniResult<JObject<'a>> {
+    let description = JObject::from(env.new_string(EXPECTED_DESCRIPTION)?);
+    env.new_object(
+        ARITHMETIC_EXCEPTION_CLASS,
+        "(Ljava/lang/String;)V",
+        &[description.into()],
+    )
+}
+
+fn assert_jni_call_default_errors<F, R>(executor: &Executor, expected_error: ErrorMatch, closure: F)
+where
+    F: FnOnce(&JNIEnv) -> JniResult<R>,
+    R: std::fmt::Debug,
+{
+    let result = jni_call_default(executor, closure);
+    assert_result_is_expected_error(expected_error, result);
+}
+
+fn assert_jni_call_transaction_errors<F, R>(
+    executor: &Executor,
+    expected_error: ErrorMatch,
+    closure: F,
+) where
+    F: FnOnce(&JNIEnv) -> JniResult<R>,
+    R: std::fmt::Debug,
+{
+    let result = jni_call_transaction(executor, closure);
+    assert_result_is_expected_error(expected_error, result);
+}
+
+fn assert_result_is_expected_error<R>(expected_error: ErrorMatch, result: Result<R, ExecutionError>)
+where
+    R: std::fmt::Debug,
+{
+    assert!(result.is_err());
+    let err = result.unwrap_err();
+    assert_eq!(err, expected_error);
+}
+
+/// Performs some trivial action using JNI to wait for Java exception being thrown.
+fn wait_for_thrown_exception<'a>(env: &'a JNIEnv) -> JniResult<JObject<'a>> {
```

Comment:
It does not wait for exception (it is already thrown after successful #throw), it invokes an exception-unsafe operation to trigger a JNI error of type 'Exception' which is caused by a previously thrown exception. 

I'd clarify the docs:
> Performs an exception-unsafe JNI operation to trigger a JNI error caused by an already thrown Java exception.

Also, I'd check the precondition that there is an exception thrown (assert(env.exception_check)).